### ISSUE TITLE
Maximum recursion depth exceeded in __getattr__().

### ISSUE THREAD

**Issue by felixxm** [python/cpython#103272]
# Bug report

We're hitting `RecursionError: maximum recursion depth exceeded` in Django test suite with Python 3.12.0a7 when accessing an attribute with a custom `__getattr__()` method:
```
ERROR: test_access_warning (deprecation.test_storages.DefaultStorageDeprecationTests.test_access_warning)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/django/tests/deprecation/test_storages.py", line 128, in test_access_warning
    settings.DEFAULT_FILE_STORAGE
  File "/django/django/conf/__init__.py", line 83, in __getattr__
    if (_wrapped := self._wrapped) is empty:
                    ^^^^^^^^^^^^^
  File "/django/django/conf/__init__.py", line 83, in __getattr__
    if (_wrapped := self._wrapped) is empty:
                    ^^^^^^^^^^^^^
  File "/django/django/conf/__init__.py", line 83, in __getattr__
    if (_wrapped := self._wrapped) is empty:
                    ^^^^^^^^^^^^^
  [Previous line repeated 790 more times]
RecursionError: maximum recursion depth exceeded
```

See [affected test](https://github.com/django/django/blob/fdf0a367bdd72c70f91fb3aed77dabbe9dcef69f/tests/deprecation/test_storages.py#L47-L49) and [`LazySettings.__getattr__()`](https://github.com/django/django/blob/fdf0a367bdd72c70f91fb3aed77dabbe9dcef69f/django/conf/__init__.py#L81-L96).

Bisected to the aa0a73d1bc53dcb6348a869df1e775138991e561.

I'd try to prepare a small regression test.

# Your environment

- CPython versions tested on: Python 3.12.0a7
- Operating system and architecture: x86_64 GNU/Linux

### Linked PRs
* gh-103336
* gh-103351

**Comment by AlexWaygood** [comment]
> Bisected to the [aa0a73d](https://github.com/python/cpython/commit/aa0a73d1bc53dcb6348a869df1e775138991e561).

Cc. @wangxiang-hz and @Fidget-Spinner

**Comment by Fidget-Spinner** [comment]
Thanks for finding this. I'll try to debug once there's a minimal reproducer. Sorry I'm a little swamped for the next 2 weeks.

**Comment by sunmy2019** [comment]
I got a minimal repo!

```py 
class A:
    def __init__(self) -> None:
        self.bar = 0

    def __getattribute__(self, name):
        return super().__getattribute__(name)

    def __getattr__(self, name):
        if self.bar == 0:
            raise ValueError

    @property
    def foo(self):
        return self.__getattr__("foo")


A().foo
```

Running in older Python version, we got

```py
Traceback (most recent call last):
  File "/home/ubuntu/django-main/tests/../a.py", line 17, in <module>
    A().foo
  File "/home/ubuntu/django-main/tests/../a.py", line 6, in __getattribute__
    return super().__getattribute__(name)
  File "/home/ubuntu/django-main/tests/../a.py", line 14, in foo
    return self.__getattr__("foo")
  File "/home/ubuntu/django-main/tests/../a.py", line 10, in __getattr__
    raise ValueError
ValueError
```

Call flow:
```
__getattribute__ called with name='foo'
property `foo` called

__getattribute__ called with name='__getattr__'
__getattr__ called with name='foo'
__getattribute__ called with name='bar'
```

But with the new python in main branch, (with out pydebug)

```py
Exception ignored in tp_clear of: <class 'type'>
Traceback (most recent call last):
  File "/home/ubuntu/django-main/tests/../a.py", line 6, in __getattribute__
    return super().__getattribute__(name)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
SystemError: <method-wrapper '__getattribute__' of A object at 0x7fc8af29e090> returned a result with an exception set
Traceback (most recent call last):
  File "/home/ubuntu/django-main/tests/../a.py", line 17, in <module>
    A().foo
  File "/home/ubuntu/django-main/tests/../a.py", line 9, in __getattr__
    if self.bar == 0:
       ^^^^^^^^
  File "/home/ubuntu/django-main/tests/../a.py", line 9, in __getattr__
    if self.bar == 0:
       ^^^^^^^^
  File "/home/ubuntu/django-main/tests/../a.py", line 9, in __getattr__
    if self.bar == 0:
       ^^^^^^^^
  [Previous line repeated 71 more times]
  File "/home/ubuntu/django-main/tests/../a.py", line 10, in __getattr__
    raise ValueError
ValueError
```

More precuriously, when I try to add some debug lines like this:
```py
class A:
    def __init__(self) -> None:
        self.bar = 0

    def __getattribute__(self, name):
        return super().__getattribute__(name)

    def __getattr__(self, name):
        print(name)   ######### <----------- Here
        if self.bar == 0:
            raise ValueError

    @property
    def foo(self):
        return self.__getattr__("foo")


A().foo
```

The new python outputs,
```py
foo
Traceback (most recent call last):
  File "/home/ubuntu/django-main/tests/../a.py", line 18, in <module>
    A().foo
  File "/home/ubuntu/django-main/tests/../a.py", line 9, in __getattr__
    print(name)
  File "/home/ubuntu/django-main/tests/../a.py", line 6, in __getattribute__
    return super().__getattribute__(name)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ubuntu/django-main/tests/../a.py", line 15, in foo
    return self.__getattr__("foo")
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ubuntu/django-main/tests/../a.py", line 11, in __getattr__
    raise ValueError
ValueError
```

Everything just works fine? Looks like something in C is corrupting ❓ ❓ ❓

**Comment by sunmy2019** [comment]
Haha, now with `pydebug` on, I got the output from new python:
```
foo
python: Python/ceval.c:821: _PyEval_EvalFrameDefault: Assertion `!_PyErr_Occurred(tstate)' failed.
Aborted (core dumped)
```

**Comment by hauntsaninja** [comment]
Thanks for the report and for narrowing this down!

The original change has been reverted and we've added a regression test, so I think this can be closed

**Comment by zzzeek** [comment]
WHEW, hi all, been bisecting all day (including getting better at bisecting large C code bases) to get here.

SQLAlchemy has been having failures on this in 3.12.0a7 and seem to be good in 3.12.0b1, just wanted to make sure this was logged.

FTR, here's our test case:

```py
import sys


class Thing:
    @property
    def comparator(self):
        raise TypeError("a type error")

    def __getattr__(self, name):
        return getattr(self.comparator, name)

t1 = Thing()

try:
    t1.comparator
except TypeError:
    print("got typeerror as expected")

```

with the bug it throws a RecursionError

**Comment by wangxiang-hz** [comment]
> 3.12.0a7

 I have confirmed that version 3.12.0a7 was released on April 4th and the bug does exist in this version. However, the bug has been fixed in version 3.12.0b1. Therefore, it is highly likely that your error is related to this issue.

**Comment by sunmy2019** [comment]
> your error is related to this issue

It is. Test coverage is added in #103272
